File: reference/SigmoidNeuron.m
classdef SigmoidNeuron < hgsetget
	properties
        inweights
        bias
    end
    methods
        function obj = SigmoidNeuron(w, b)
            obj.inweights = w;
            obj.bias = b;
        end
        function r = calcSigmoidNeuron(nn, in)
            if length(in) ~= length(nn.inweights)
                error('In values do not match weights');
            end
            z = -nn.bias;
            for i = 1:length(in)
                z = z + in(i)*nn.inweights(i);
            end
            r = 1/(1+(exp(1)^(-3.1*z)));
        end
        function n = mateSigmoidNeuron(a, b, threshold)
            if nargin < 3
                threshold = 0.51337;
            end
            if length(a.inweights) ~= length(b.inweights)
                error('Incompatible parents');
            end
            n = SigmoidNeuron(a.inweights, a.bias);
            for i = 1:length(a.inweights)
                if rand() > threshold
                     n.inweights(i) = b.inweights(i);
                end
            end
            if rand() > threshold
                n.bias = b.bias;
            end
        end
        function m = mutateSigmoidNeuron(a, threshold, maxweight, maxbias)
            m = SigmoidNeuron(a.inweights, a.bias);
            for i = 1:length(a.inweights)
                if rand() > threshold
                    m.inweights(i) = floor(mod(rand()*(10*ceil(log10(maxweight+1))), maxweight+1));
                end
            end
            if rand() > threshold
                m.bias = floor(mod(rand()*(10*ceil(log10(maxbias+1))), maxbias+1));
            end
        end
    end
end

File: src/SigmoidNeuron.c
#include "SigmoidNeuron.h"

SigmoidNeuron SigmoidNeuron_new(const List inweights, double bias) {
    SigmoidNeuron sn = calloc(sizeof(SigmoidNeuron_nonptr), 1);
    sn->inweights = List_dup(inweights);
    sn->bias = bias;
    return sn;
}

void *_SigmoidNeuron_copier(void *a) {
    SigmoidNeuron n = calloc(sizeof(SigmoidNeuron_nonptr), 1);
    n->inweights = List_dup(((SigmoidNeuron)a)->inweights);
    n->bias = ((SigmoidNeuron)a)->bias;
    return n;
}

double SigmoidNeuron_calc(const SigmoidNeuron sn, const double *indata, int indatalen) {
    double z, r; int i; ListElem curr;
    if (indatalen != List_length(sn->inweights)) {
        error("In values do not match number of weights");
    }
    z = -sn->bias;
    curr = List_startElem(sn->inweights);
    for (i = 0; i < indatalen; i++) {
        z += indata[i] * *((double *)ListElem_value(curr));
        curr = ListElem_next(curr);
    }
    r = 1/(1+pow(M_E, -3.1*z));
    return r;
}

/* mating sigmoid neuron has 0 relation to inputs */
SigmoidNeuron SigmoidNeuron_mate(const SigmoidNeuron a, const SigmoidNeuron b, double threshold) {
    int i; SigmoidNeuron n; ListElem currn, currb;
    if (threshold < 0) threshold = 0.51337;
    if (List_length(a->inweights) != List_length(b->inweights)) {
        error("Incompatible parents");
    }
    /* completely new copy of a */
    n = SigmoidNeuron_new(List_copy(a->inweights, &_Double_copier), a->bias);
    currn = List_startElem(n->inweights);
    currb = List_startElem(b->inweights);
    for (i = 0; i < List_length(a->inweights); i++) {
        if (randdouble() > threshold) {
            /* replace the old value */
            free(ListElem_value(currn));
            ListElem_value(currn) = _Double_new(*(double *)ListElem_value(currb));
        }
        currn = ListElem_next(currn);
        currb = ListElem_next(currb);
    }
    if (randdouble() > threshold) {
        n->bias = b->bias;
    }
    return n;
}

/* mutate does not rely on old values at all! can free() old neuron after calling */
SigmoidNeuron SigmoidNeuron_mutate(const SigmoidNeuron a, double threshold, double maxweight, double maxbias) {
    int i; SigmoidNeuron m; ListElem curra, currm;
    m = SigmoidNeuron_new(a->inweights, a->bias);
    curra = List_startElem(a->inweights);
    currm = List_startElem(m->inweights);
    for (i = 0; i < List_length(a->inweights); i++) {
        if (randdouble() > threshold) {
            ListElem_value(currm) = _Double_new(randWithMaxDouble(maxweight)); /* not evenly distributed, whoops TODO fix this */
        } else {
            ListElem_value(currm) = _Double_new(*(double *)ListElem_value(curra));
        }
        curra = ListElem_next(curra);
        currm = ListElem_next(currm);
    }
    if (randdouble() > threshold) {
        m->bias = randWithMaxDouble(maxbias);
    }
    return m;
}

void SigmoidNeuron_delete(SigmoidNeuron sn) {
    List_delete(sn->inweights, &free);
    free(sn);
}
